File: h2o-algos/src/main/java/water/userapi/GradientBoosting.java
package water.userapi;

import hex.ModelMetricsRegression;
import hex.tree.gbm.GBM;
import hex.tree.gbm.GBMModel;
import water.Scope;

/**
 * UserApi adapter for GBM operations
 * 
 * Created by vpatryshev on 2/27/17.
 */
class GradientBoosting {
  
  /*private*/ final GBMModel.GBMParameters params = new GBMModel.GBMParameters();
  
  GradientBoosting(Dataset.TrainAndValid split) {
    params._train = split.train.frame()._key;
    params._valid = split.valid.frame()._key;
  }

  public volatile GBM trainingJob = null;

  public GBM trainingJob() {
    if (trainingJob == null) trainingJob = new GBM(params);
    return trainingJob;
  }

  volatile private GBMModel model = null;

  // Build a first model; all remaining models should be equal
  private GBMModel model() {
    if (model == null) model = Scope.track(trainingJob().trainModel().get());
    return model;
  }
  
  public ModelMetricsRegression metrics() {
    return (ModelMetricsRegression) model()._output._validation_metrics;
  }

  public GradientBoosting responseColumn(String rc) {
    params._response_column = rc;
    return this;
  }

  public GradientBoosting seed(int i) {
    params._seed = i;
    return this;
  }

  public GradientBoosting minRows(int i) {
    params._min_rows = i;
    return this;
  }

  public GradientBoosting maxDepth(int i) {
    params._max_depth = i;
    return this;
  }

  public GradientBoosting nTrees(int i) {
    params._ntrees = i;
    return this;
  }

  public GradientBoosting colSampleRate(double colSampleRate) {
    params._col_sample_rate = colSampleRate;
    return this;
  }

  public GradientBoosting colSampleRatePerTree(double colSampleRatePerTree) {
    params._col_sample_rate_per_tree = colSampleRatePerTree;
    return this;
  }

  public GradientBoosting sampleRate(double sampleRate) {
    params._sample_rate = sampleRate;
    return this;
  }

  public GradientBoosting learnRate(float v) {
    params._learn_rate = v;
    return this;
  }

  public int nTrees() {
    return params._ntrees;
  }

  public GradientBoosting scoreTreeInterval(int i) {
    params._score_tree_interval = i;
    return this;
  }
}

File: h2o-scala/src/main/scala/water/userapi/GradientBoosting.scala
package water.userapi

import hex.ModelMetricsRegression
import hex.tree.gbm.GBM
import hex.tree.gbm.GBMModel
import water.Scope
import water.fvec.Frame

/**
  * UserApi adapter for GBM operations
  *
  * Created by vpatryshev on 2/27/17.
  */
class GradientBoosting {

  val params = new GBMModel.GBMParameters()

  def this(train: Frame, valid: Frame) {
    this()
    params._train = train._key
    params._valid = valid._key
  }

  @volatile lazy val trainingJob: GBM = new GBM(params)

  @volatile lazy val model: GBMModel = Scope.track(trainingJob.trainModel.get)

  def output: GBMModel.GBMOutput = model._output

  def metrics: Option[ModelMetricsRegression] = {
    Option(output._validation_metrics) collect {
      case mmr: ModelMetricsRegression => mmr
    }
  }

  def responseColumn(rc: String): GradientBoosting = {
    params._response_column = rc
    this
  }

  def seed(i: Int): GradientBoosting ={
    params._seed = i
    this
  }

  def minRows(i: Int): GradientBoosting ={
    params._min_rows = i
    this
  }

  def maxDepth(i: Int): GradientBoosting ={
    params._max_depth = i
    this
  }

  def nTrees(i: Int): GradientBoosting ={
    params._ntrees = i
    this
  }

  def colSampleRate(colSampleRate: Double): GradientBoosting = {
    params._col_sample_rate = colSampleRate
    this
  }

  def colSampleRatePerTree(colSampleRatePerTree: Double): GradientBoosting = {
    params._col_sample_rate_per_tree = colSampleRatePerTree
    this
  }

  def sampleRate(sampleRate: Double): GradientBoosting = {
    params._sample_rate = sampleRate
    this
  }

  def learnRate(v: Float): GradientBoosting = {
    params._learn_rate = v
    this
  }

  def nTrees: Int = params._ntrees

  def scoreTreeInterval(i: Int): GradientBoosting = {
    params._score_tree_interval = i
    this
  }

  def minSplitImprovement(v: Double): GradientBoosting = {
    params._min_split_improvement = v
    this
  }

  def logLoss: Double = 
    output._scored_valid(output._scored_valid.length - 1)._logloss
}